### PR TITLE
Feat/vogue upload micro

### PR DESCRIPTION
**Pull request by Mropat**
This PR adds/fixes ...
Upload microsalt results to Vogue
Added:
`cg workflow microsalt upload-analysis-vogue <case_id>` 
Uploads latest trending data from latest analysis for given case
`cg workflow microsalt upload-latest-analyses-vogue`
Uploads latest trending data from latest analyses for all cases where un-uploaded microbial analysis trending data exists

TODOs:
Tests will be added shortly (if needed)

### How to prepare for test
- [x] ssh to hasta
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [x] do `cg workflow microsalt upload-analysis-vogue bigdrum`
- [x] do `cg workflow microsalt upload-latest-analyses-vogue --dry-run`

### Expected test outcome
- [x] check that  `upload-analysis-vogue` command executes successfully
- [x] check that `upload-latest-analyses-vogue` command finds latest analyses and informes what it would have uploaded 
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [x] code approved by @mayabrandi 
- [x] tests executed by MR
- [x] "Merge and deploy" approved by MR
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan
- [ ] Document in Atlas when we're allowed to add things to Atlas again
- [ ] Deploy this branch
- [ ] Inform @hassanfa and production about trended vogue data from microsalt
- [ ] Create cronjob for automatic upload of microsalt to vogue

### CODE REVIEW

**Comment by Mropat** [comment]
Testing store new analysis in housekeeper

```
[0|0|1102] [maryia.ropart@hasta:~] [S_main] 4s $ cg workflow microsalt store luckyhog
Found deliverables file /home/proj/stage/microbial/results/reports/deliverables/939846_deliverables.yaml
Converting pipeline deliverables to CG deliverables
Running command /home/proj/bin/conda/envs/S_hermes/bin/hermes convert deliverables --pipeline microsalt /home/proj/stage/microbial/results/reports/deliverables/939846_deliverables.yaml
Fetching bundle with name: luckyhog
Created new bundle: luckyhog
Created new version
Use global root path /home/proj/stage/housekeeper-bundles
created new bundle version dir: /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/sampleinfo.json -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/sampleinfo.json
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/939846_QC_2.html -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/939846_QC_2.html
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/939846_Typing_2.html -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/939846_Typing_2.html
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/ACC6513.json -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513.json
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/config.log -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/config.log
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/assembly/ACC6513A116_trimmed_contigs.fasta -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_trimmed_contigs.fasta
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/trimmed/ACC6513A116_trim_front_pair.fastq.gz -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_trim_front_pair.fastq.gz
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/trimmed/ACC6513A116_trim_rev_pair.fastq.gz -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_trim_rev_pair.fastq.gz
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/trimmed/ACC6513A116_trim_unpair.fastq.gz -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_trim_unpair.fastq.gz
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/slurm_ACC6513A116.log -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/slurm_ACC6513A116.log
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/assembly/quast/ACC6513A116_report.tsv -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_report.tsv
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/alignment/ACC6513A116_NZ_CP014016.2.bam_sort -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_NZ_CP014016.2.bam_sort
linked file: /home/proj/stage/microbial/results/ACC6513A116_2020.12.30_11.45.26/alignment/ACC6513A116_NZ_CP014016.2.stats.ins -> /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513A116_NZ_CP014016.2.stats.ins
Analysis successfully stored in Housekeeper: luckyhog : 2020-12-30 12:04:54
Found deliverables file /home/proj/stage/microbial/results/reports/deliverables/939846_deliverables.yaml
Running command source activate S_microSALT; /home/proj/bin/conda/envs/S_microSALT/bin/microSALT --version
Analysis successfully stored in StatusDB: luckyhog : 2020-12-30 12:04:54
Action None set for case luckyhog
```

Testing upload trending data to vogue
```
[0|0|1094] [maryia.ropart@hasta:~] [S_main] 42s $ cg workflow microsalt upload-analysis-vogue luckyhog
Running command source activate S_microSALT; /home/proj/bin/conda/envs/S_microSALT/bin/microSALT --version
Fetching files from bundle luckyhog
Fetching files with tags in [vogue]
Running command /home/proj/bin/conda/envs/S_vogue/bin/vogue --config /home/proj/stage/servers/config/hasta.scilifelab.se/vogue-stage.yaml load bioinfo raw --sample-list ACC6513A95,ACC6513A96,ACC6513A97,ACC6513A98,ACC6513A99,ACC6513A100,ACC6513A101,ACC6513A102,ACC6513A103,ACC6513A104,ACC6513A105,ACC6513A106,ACC6513A107,ACC6513A108,ACC6513A109,ACC6513A110,ACC6513A111,ACC6513A112,ACC6513A113,ACC6513A114,ACC6513A115,ACC6513A116,ACC6513A117,ACC6513A118,ACC6513A119,ACC6513A120,ACC6513A121,ACC6513A122,ACC6513A123,ACC6513A124,ACC6513A125,ACC6513A126,ACC6513A127,ACC6513A128,ACC6513A129,ACC6513A130,ACC6513A131,ACC6513A132,ACC6513A133,ACC6513A134,ACC6513A135,ACC6513A136,ACC6513A137,ACC6513A138,ACC6513A139,ACC6513A140,ACC6513A141,ACC6513A142,ACC6513A143,ACC6513A144,ACC6513A145,ACC6513A146,ACC6513A147,ACC6513A148,ACC6513A149,ACC6513A150,ACC6513A151,ACC6513A152,ACC6513A153,ACC6513A154,ACC6513A155,ACC6513A156,ACC6513A157,ACC6513A158,ACC6513A159,ACC6513A160,ACC6513A161,ACC6513A162,ACC6513A163,ACC6513A164,ACC6513A165,ACC6513A166,ACC6513A167,ACC6513A168,ACC6513A169,ACC6513A170,ACC6513A171,ACC6513A172,ACC6513A173,ACC6513A174,ACC6513A175,ACC6513A176,ACC6513A233,ACC6513A234,ACC6513A235,ACC6513A236,ACC6513A237,ACC6513A238,ACC6513A239,ACC6513A240,ACC6513A241,ACC6513A242,ACC6513A243,ACC6513A244,ACC6513A245,ACC6513A246,ACC6513A247,ACC6513A248,ACC6513A249,ACC6513A250,ACC6513A251,ACC6513A252,ACC6513A253,ACC6513A254,ACC6513A255,ACC6513A256,ACC6513A1,ACC6513A2,ACC6513A3,ACC6513A4,ACC6513A5,ACC6513A6,ACC6513A7,ACC6513A8,ACC6513A9,ACC6513A10,ACC6513A11,ACC6513A12,ACC6513A13,ACC6513A14,ACC6513A15,ACC6513A16,ACC6513A17,ACC6513A18,ACC6513A19,ACC6513A20,ACC6513A21,ACC6513A22,ACC6513A23,ACC6513A24,ACC6513A25,ACC6513A26,ACC6513A27,ACC6513A28,ACC6513A29,ACC6513A30,ACC6513A31,ACC6513A32,ACC6513A33,ACC6513A34,ACC6513A35,ACC6513A36,ACC6513A37,ACC6513A38,ACC6513A39,ACC6513A40,ACC6513A41,ACC6513A42,ACC6513A43,ACC6513A44,ACC6513A45,ACC6513A46,ACC6513A47,ACC6513A48,ACC6513A49,ACC6513A50,ACC6513A51,ACC6513A52,ACC6513A53,ACC6513A54,ACC6513A55,ACC6513A56,ACC6513A57,ACC6513A58,ACC6513A59,ACC6513A60,ACC6513A61,ACC6513A62,ACC6513A63,ACC6513A64,ACC6513A65,ACC6513A66,ACC6513A67,ACC6513A68,ACC6513A69,ACC6513A70,ACC6513A71,ACC6513A72,ACC6513A73,ACC6513A74,ACC6513A75,ACC6513A76,ACC6513A77,ACC6513A78,ACC6513A79,ACC6513A80,ACC6513A81,ACC6513A82,ACC6513A83,ACC6513A84,ACC6513A85,ACC6513A86,ACC6513A87,ACC6513A88,ACC6513A89,ACC6513A90,ACC6513A91,ACC6513A92,ACC6513A93,ACC6513A94,ACC6513A177,ACC6513A178,ACC6513A179,ACC6513A180,ACC6513A181,ACC6513A182,ACC6513A183,ACC6513A184,ACC6513A185,ACC6513A186,ACC6513A187,ACC6513A188,ACC6513A189,ACC6513A190,ACC6513A191,ACC6513A192,ACC6513A193,ACC6513A194,ACC6513A195,ACC6513A196,ACC6513A197,ACC6513A198,ACC6513A199,ACC6513A200,ACC6513A201,ACC6513A202,ACC6513A203,ACC6513A204,ACC6513A205,ACC6513A206,ACC6513A207,ACC6513A208,ACC6513A209,ACC6513A210,ACC6513A211,ACC6513A212,ACC6513A213,ACC6513A214,ACC6513A215,ACC6513A216,ACC6513A217,ACC6513A218,ACC6513A219,ACC6513A220,ACC6513A221,ACC6513A222,ACC6513A223,ACC6513A224,ACC6513A225,ACC6513A226,ACC6513A227,ACC6513A228,ACC6513A229,ACC6513A230,ACC6513A231,ACC6513A232 --analysis-result /home/proj/stage/housekeeper-bundles/luckyhog/2020-12-30/ACC6513.json --analysis-type microsalt --analysis-case luckyhog --workflow-version 3.1.3 --case-analysis-type microsalt --analysis-workflow microsalt
Running command /home/proj/bin/conda/envs/S_vogue/bin/vogue --config /home/proj/stage/servers/config/hasta.scilifelab.se/vogue-stage.yaml load bioinfo process --analysis-type microsalt --analysis-case luckyhog --analysis-workflow microsalt --workflow-version 3.1.3 --case-analysis-type microsalt
Running command /home/proj/bin/conda/envs/S_vogue/bin/vogue --config /home/proj/stage/servers/config/hasta.scilifelab.se/vogue-stage.yaml load bioinfo sample --analysis-case luckyhog
Successfully uploaded latest analysis data for case luckyhog to Vogue!

```

**Comment by Mropat** [comment]
@mayabrandi @patrikgrenfeldt  ready for review

**Review by mayabrandi — APPROVED** [review]
Code wise this looks fine Maria. I have nothing to complain about. I was hesitant about the fact that you were adding the load of vogue as part of the micro workflow but since this is a quick and dirty fix its okej. Well decouple the Vogue load cli from cg later.

File: cg/apps/vogue.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def __init__(self, config: dict):
         self.vogue_binary = config["vogue"]["binary_path"]
         self.process = Process(binary=self.vogue_binary, config=self.vogue_config)
 
-    def load_genotype_data(self, genotype_dict: dict):
+    def load_genotype_data(self, genotype_dict: dict) -> None:
```
Comment (moonso) [inline cg/apps/vogue.py:24]: mmm `dict` 😆

File: cg/cli/upload/vogue.py [hunk 2]
```diff
@@ -236,9 +236,7 @@ def _get_samples(store: Store, case_name: str) -> str:
     """
 
     link_objs = get_links(store, case_name)
-    sample_ids = set()
-    for link_obj in link_objs:
-        sample_ids.add(link_obj.sample.internal_id)
+    sample_ids = {link_obj.sample.internal_id for link_obj in link_objs}
```
Comment (moonso) [inline cg/cli/upload/vogue.py:239]: Hello sourcery!

File: cg/cli/workflow/microsalt/base.py [hunk 3]
```diff
@@ -260,3 +266,82 @@ def start_available(context: click.Context, dry_run: bool):
...
+    """Upload the trending report for latest analysis of given case_id to Vogue"""
+
+    vogue_api = VogueAPI(context.obj)
+    microsalt_analysis_api: MicrosaltAnalysisAPI = context.obj["microsalt_analysis_api"]
+    case_obj = microsalt_analysis_api.db.family(unique_id)
+    if not case_obj or not case_obj.analyses:
+        LOG.error("No analysis available for %s", unique_id)
+        raise click.Abort
+
+    samples_string = ",".join(str(link_obj.sample.internal_id) for link_obj in case_obj.links)
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:285]:
```suggestion
    samples_string: List[str] = ",".join(str(link_obj.sample.internal_id) for link_obj in case_obj.links)
```

File: cg/cli/workflow/microsalt/base.py [hunk 4]
```diff
@@ -260,3 +266,82 @@ def start_available(context: click.Context, dry_run: bool):
...
+    microsalt_version = microsalt_analysis_api.get_microsalt_version()
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:286]:
```suggestion
    microsalt_versio: float = microsalt_analysis_api.get_microsalt_version()
```
?

File: cg/cli/workflow/microsalt/base.py [hunk 5]
```diff
@@ -260,3 +266,82 @@ def start_available(context: click.Context, dry_run: bool):
...
+        "analysis_type": "microsalt",
+        "analysis_workflow_name": "microsalt",
+        "analysis_workflow_version": microsalt_version,
+        "case_analysis_type": "microsalt",
+    }
+    vogue_api.load_bioinfo_raw(load_bioinfo_inputs=vogue_load_args)
+    vogue_api.load_bioinfo_process(load_bioinfo_inputs=vogue_load_args, cleanup_flag=False)
+    vogue_api.load_bioinfo_sample(load_bioinfo_inputs=vogue_load_args)
+    case_obj.analyses[0].uploaded_at = dt.datetime.now()
+    microsalt_analysis_api.db.commit()
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:317]: Hehe, little bit weird but works

File: cg/cli/workflow/microsalt/base.py [hunk 6]
```diff
@@ -260,3 +266,82 @@ def start_available(context: click.Context, dry_run: bool):
...
+
+@microsalt.command("upload-latest-analyses-vogue")
+@OPTION_DRY_RUN
+@click.pass_context
+def upload_vogue_latest(context: click.Context, dry_run: bool) -> None:
+    """Upload the trending reports for all un-uploaded latest analyses to Vogue"""
+
+    EXIT_CODE: int = EXIT_SUCCESS
+    microsalt_analysis_api: MicrosaltAnalysisAPI = context.obj["microsalt_analysis_api"]
+    latest_analyses = list(
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:329]: Could not that function return None?

File: cg/meta/workflow/microsalt.py [hunk 7]
```diff
@@ -453,3 +455,11 @@ def submit_trailblazer_analysis(self, case_id: str) -> None:
             priority=self.get_priority(case_id),
             data_analysis=Pipeline.MICROSALT,
         )
+
+    def get_microsalt_version(self):
```
Comment (moonso) [inline cg/meta/workflow/microsalt.py:459]: Aha string!
```suggestion
    def get_microsalt_version(self) -> str:
```

**Review by moonso — APPROVED** [review]
Great, some minor comments

File: cg/cli/upload/vogue.py (hunk 2 above)
Comment (Mropat) [reply cg/cli/upload/vogue.py:239]: :yum:

File: cg/cli/workflow/microsalt/base.py (hunk 5 above)
Comment (Mropat) [reply cg/cli/workflow/microsalt/base.py:317]: :woman_shrugging:

File: cg/cli/workflow/microsalt/base.py (hunk 6 above)
Comment (Mropat) [reply cg/cli/workflow/microsalt/base.py:329]: its a query, which can be empty, and then it turns into empty list

File: cg/cli/workflow/microsalt/base.py (hunk 4 above)
Comment (Mropat) [reply cg/cli/workflow/microsalt/base.py:286]: its a string! because its formatted as "0.0.1"

**Comment by Mropat** [comment]
Deployed on Prod

```
Building wheels for collected packages: cg
  Running setup.py bdist_wheel for cg ... done
  Stored in directory: /home/maryia.ropart/tmp/pip-ephem-wheel-cache-df1rwi4z/wheels/5c/23/37/e3635140fad58f81760c6db5705a64683daf19230a7bb26dda
Successfully built cg
Installing collected packages: cg
  Found existing installation: cg 18.7.2
    Uninstalling cg-18.7.2:
      Successfully uninstalled cg-18.7.2
Successfully installed cg-18.8.0

```

**Pull request merged**